File: src/libqcolorspace.zig
const C = @import("qt6c");
const qcolorspace_enums = enums;
const std = @import("std");

/// https://doc.qt.io/qt-6/qcolorspace.html
pub const qcolorspace = struct {
    /// New constructs a new QColorSpace object.
    ///
    ///
    pub fn New() ?*C.QColorSpace {
        return C.QColorSpace_new();
    }

    /// New2 constructs a new QColorSpace object.
    ///
    /// ``` namedColorSpace: qcolorspace_enums.NamedColorSpace ```
    pub fn New2(namedColorSpace: i64) ?*C.QColorSpace {
        return C.QColorSpace_new2(@intCast(namedColorSpace));
    }

    /// New3 constructs a new QColorSpace object.
    ///
    /// ``` primaries: qcolorspace_enums.Primaries, transferFunction: qcolorspace_enums.TransferFunction ```
    pub fn New3(primaries: i64, transferFunction: i64) ?*C.QColorSpace {
        return C.QColorSpace_new3(@intCast(primaries), @intCast(transferFunction));
    }

    /// New4 constructs a new QColorSpace object.
    ///
    /// ``` primaries: qcolorspace_enums.Primaries, gamma: f32 ```
    pub fn New4(primaries: i64, gamma: f32) ?*C.QColorSpace {
        return C.QColorSpace_new4(@intCast(primaries), @floatCast(gamma));
    }

    /// New5 constructs a new QColorSpace object.
    ///
    /// ``` primaries: qcolorspace_enums.Primaries, transferFunctionTable: []u16 ```
    pub fn New5(primaries: i64, transferFunctionTable: []u16) ?*C.QColorSpace {
        const transferFunctionTable_list = C.struct_libqt_list{
            .len = transferFunctionTable.len,
            .data = transferFunctionTable.ptr,
        };

        return C.QColorSpace_new5(@intCast(primaries), transferFunctionTable_list);
    }

    /// New6 constructs a new QColorSpace object.
    ///
    /// ``` whitePoint: ?*C.QPointF, redPoint: ?*C.QPointF, greenPoint: ?*C.QPointF, bluePoint: ?*C.QPointF, transferFunction: qcolorspace_enums.TransferFunction ```
    pub fn New6(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, transferFunction: i64) ?*C.QColorSpace {
        return C.QColorSpace_new6(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), @intCast(transferFunction));
    }

    /// New7 constructs a new QColorSpace object.
    ///
    /// ``` whitePoint: ?*C.QPointF, redPoint: ?*C.QPointF, greenPoint: ?*C.QPointF, bluePoint: ?*C.QPointF, transferFunctionTable: []u16 ```
    pub fn New7(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, transferFunctionTable: []u16) ?*C.QColorSpace {
        const transferFunctionTable_list = C.struct_libqt_list{
            .len = transferFunctionTable.len,
            .data = transferFunctionTable.ptr,
        };

        return C.QColorSpace_new7(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), transferFunctionTable_list);
    }

    /// New8 constructs a new QColorSpace object.
    ///
    /// ``` whitePoint: ?*C.QPointF, redPoint: ?*C.QPointF, greenPoint: ?*C.QPointF, bluePoint: ?*C.QPointF, redTransferFunctionTable: []u16, greenTransferFunctionTable: []u16, blueTransferFunctionTable: []u16 ```
    pub fn New8(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, redTransferFunctionTable: []u16, greenTransferFunctionTable: []u16, blueTransferFunctionTable: []u16) ?*C.QColorSpace {
        const redTransferFunctionTable_list = C.struct_libqt_list{
            .len = redTransferFunctionTable.len,
            .data = redTransferFunctionTable.ptr,
        };
        const greenTransferFunctionTable_list = C.struct_libqt_list{
            .len = greenTransferFunctionTable.len,
            .data = greenTransferFunctionTable.ptr,
        };
        const blueTransferFunctionTable_list = C.struct_libqt_list{
            .len = blueTransferFunctionTable.len,
            .data = blueTransferFunctionTable.ptr,
        };

        return C.QColorSpace_new8(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), redTransferFunctionTable_list, greenTransferFunctionTable_list, blueTransferFunctionTable_list);
    }

    /// New9 constructs a new QColorSpace object.
    ///
    /// ``` colorSpace: ?*C.QColorSpace ```
    pub fn New9(colorSpace: ?*anyopaque) ?*C.QColorSpace {
        return C.QColorSpace_new9(@ptrCast(colorSpace));
    }

    /// New10 constructs a new QColorSpace object.
    ///
    /// ``` primaries: qcolorspace_enums.Primaries, transferFunction: qcolorspace_enums.TransferFunction, gamma: f32 ```
    pub fn New10(primaries: i64, transferFunction: i64, gamma: f32) ?*C.QColorSpace {
        return C.QColorSpace_new10(@intCast(primaries), @intCast(transferFunction), @floatCast(gamma));
    }

    /// New11 constructs a new QColorSpace object.
    ///
    /// ``` whitePoint: ?*C.QPointF, redPoint: ?*C.QPointF, greenPoint: ?*C.QPointF, bluePoint: ?*C.QPointF, transferFunction: qcolorspace_enums.TransferFunction, gamma: f32 ```
    pub fn New11(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, transferFunction: i64, gamma: f32) ?*C.QColorSpace {
        return C.QColorSpace_new11(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), @intCast(transferFunction), @floatCast(gamma));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#operator=)
    ///
    /// ``` self: ?*C.QColorSpace, colorSpace: ?*C.QColorSpace ```
    pub fn OperatorAssign(self: ?*anyopaque, colorSpace: ?*anyopaque) void {
        C.QColorSpace_OperatorAssign(@ptrCast(self), @ptrCast(colorSpace));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#swap)
    ///
    /// ``` self: ?*C.QColorSpace, colorSpace: ?*C.QColorSpace ```
    pub fn Swap(self: ?*anyopaque, colorSpace: ?*anyopaque) void {
        C.QColorSpace_Swap(@ptrCast(self), @ptrCast(colorSpace));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#primaries)
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn Primaries(self: ?*anyopaque) i64 {
        return C.QColorSpace_Primaries(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#transferFunction)
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn TransferFunction(self: ?*anyopaque) i64 {
        return C.QColorSpace_TransferFunction(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#gamma)
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn Gamma(self: ?*anyopaque) f32 {
        return C.QColorSpace_Gamma(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#description)
    ///
    /// ``` self: ?*C.QColorSpace, allocator: std.mem.Allocator ```
    pub fn Description(self: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = C.QColorSpace_Description(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_str));
        const _ret = allocator.alloc(u8, _str.len) catch @panic("Memory allocation failed");
        for (0.._str.len) |_i| {
            _ret[_i] = _str.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setDescription)
    ///
    /// ``` self: ?*C.QColorSpace, description: []const u8 ```
    pub fn SetDescription(self: ?*anyopaque, description: []const u8) void {
        const description_str = C.struct_libqt_string{
            .len = description.len,
            .data = @constCast(description.ptr),
        };
        C.QColorSpace_SetDescription(@ptrCast(self), description_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setTransferFunction)
    ///
    /// ``` self: ?*C.QColorSpace, transferFunction: qcolorspace_enums.TransferFunction ```
    pub fn SetTransferFunction(self: ?*anyopaque, transferFunction: i64) void {
        C.QColorSpace_SetTransferFunction(@ptrCast(self), @intCast(transferFunction));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setTransferFunction)
    ///
    /// ``` self: ?*C.QColorSpace, transferFunctionTable: []u16 ```
    pub fn SetTransferFunctionWithTransferFunctionTable(self: ?*anyopaque, transferFunctionTable: []u16) void {
        const transferFunctionTable_list = C.struct_libqt_list{
            .len = transferFunctionTable.len,
            .data = transferFunctionTable.ptr,
        };
        C.QColorSpace_SetTransferFunctionWithTransferFunctionTable(@ptrCast(self), transferFunctionTable_list);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setTransferFunctions)
    ///
    /// ``` self: ?*C.QColorSpace, redTransferFunctionTable: []u16, greenTransferFunctionTable: []u16, blueTransferFunctionTable: []u16 ```
    pub fn SetTransferFunctions(self: ?*anyopaque, redTransferFunctionTable: []u16, greenTransferFunctionTable: []u16, blueTransferFunctionTable: []u16) void {
        const redTransferFunctionTable_list = C.struct_libqt_list{
            .len = redTransferFunctionTable.len,
            .data = redTransferFunctionTable.ptr,
        };
        const greenTransferFunctionTable_list = C.struct_libqt_list{
            .len = greenTransferFunctionTable.len,
            .data = greenTransferFunctionTable.ptr,
        };
        const blueTransferFunctionTable_list = C.struct_libqt_list{
            .len = blueTransferFunctionTable.len,
            .data = blueTransferFunctionTable.ptr,
        };
        C.QColorSpace_SetTransferFunctions(@ptrCast(self), redTransferFunctionTable_list, greenTransferFunctionTable_list, blueTransferFunctionTable_list);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#withTransferFunction)
    ///
    /// ``` self: ?*C.QColorSpace, transferFunction: qcolorspace_enums.TransferFunction ```
    pub fn WithTransferFunction(self: ?*anyopaque, transferFunction: i64) ?*C.QColorSpace {
        return C.QColorSpace_WithTransferFunction(@ptrCast(self), @intCast(transferFunction));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#withTransferFunction)
    ///
    /// ``` self: ?*C.QColorSpace, transferFunctionTable: []u16 ```
    pub fn WithTransferFunctionWithTransferFunctionTable(self: ?*anyopaque, transferFunctionTable: []u16) ?*C.QColorSpace {
        const transferFunctionTable_list = C.struct_libqt_list{
            .len = transferFunctionTable.len,
            .data = transferFunctionTable.ptr,
        };
        return C.QColorSpace_WithTransferFunctionWithTransferFunctionTable(@ptrCast(self), transferFunctionTable_list);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#withTransferFunctions)
    ///
    /// ``` self: ?*C.QColorSpace, redTransferFunctionTable: []u16, greenTransferFunctionTable: []u16, blueTransferFunctionTable: []u16 ```
    pub fn WithTransferFunctions(self: ?*anyopaque, redTransferFunctionTable: []u16, greenTransferFunctionTable: []u16, blueTransferFunctionTable: []u16) ?*C.QColorSpace {
        const redTransferFunctionTable_list = C.struct_libqt_list{
            .len = redTransferFunctionTable.len,
            .data = redTransferFunctionTable.ptr,
        };
        const greenTransferFunctionTable_list = C.struct_libqt_list{
            .len = greenTransferFunctionTable.len,
            .data = greenTransferFunctionTable.ptr,
        };
        const blueTransferFunctionTable_list = C.struct_libqt_list{
            .len = blueTransferFunctionTable.len,
            .data = blueTransferFunctionTable.ptr,
        };
        return C.QColorSpace_WithTransferFunctions(@ptrCast(self), redTransferFunctionTable_list, greenTransferFunctionTable_list, blueTransferFunctionTable_list);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setPrimaries)
    ///
    /// ``` self: ?*C.QColorSpace, primariesId: qcolorspace_enums.Primaries ```
    pub fn SetPrimaries(self: ?*anyopaque, primariesId: i64) void {
        C.QColorSpace_SetPrimaries(@ptrCast(self), @intCast(primariesId));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setPrimaries)
    ///
    /// ``` self: ?*C.QColorSpace, whitePoint: ?*C.QPointF, redPoint: ?*C.QPointF, greenPoint: ?*C.QPointF, bluePoint: ?*C.QPointF ```
    pub fn SetPrimaries2(self: ?*anyopaque, whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque) void {
        C.QColorSpace_SetPrimaries2(@ptrCast(self), @ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#detach)
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn Detach(self: ?*anyopaque) void {
        C.QColorSpace_Detach(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#isValid)
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn IsValid(self: ?*anyopaque) bool {
        return C.QColorSpace_IsValid(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#fromIccProfile)
    ///
    /// ``` iccProfile: []u8 ```
    pub fn FromIccProfile(iccProfile: []u8) ?*C.QColorSpace {
        const iccProfile_str = C.struct_libqt_string{
            .len = iccProfile.len,
            .data = @constCast(iccProfile.ptr),
        };
        return C.QColorSpace_FromIccProfile(iccProfile_str);
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#iccProfile)
    ///
    /// ``` self: ?*C.QColorSpace, allocator: std.mem.Allocator ```
    pub fn IccProfile(self: ?*anyopaque, allocator: std.mem.Allocator) []u8 {
        const _bytearray: C.struct_libqt_string = C.QColorSpace_IccProfile(@ptrCast(self));
        defer C.libqt_string_free(@constCast(&_bytearray));
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("Memory allocation failed");
        for (0.._bytearray.len) |_i| {
            _ret[_i] = _bytearray.data[_i];
        }
        return _ret;
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#transformationToColorSpace)
    ///
    /// ``` self: ?*C.QColorSpace, colorspace: ?*C.QColorSpace ```
    pub fn TransformationToColorSpace(self: ?*anyopaque, colorspace: ?*anyopaque) ?*C.QColorTransform {
        return C.QColorSpace_TransformationToColorSpace(@ptrCast(self), @ptrCast(colorspace));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#operator QVariant)
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn ToQVariant(self: ?*anyopaque) ?*C.QVariant {
        return C.QColorSpace_ToQVariant(@ptrCast(self));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#setTransferFunction)
    ///
    /// ``` self: ?*C.QColorSpace, transferFunction: qcolorspace_enums.TransferFunction, gamma: f32 ```
    pub fn SetTransferFunction2(self: ?*anyopaque, transferFunction: i64, gamma: f32) void {
        C.QColorSpace_SetTransferFunction2(@ptrCast(self), @intCast(transferFunction), @floatCast(gamma));
    }

    /// [Qt documentation](https://doc.qt.io/qt-6/qcolorspace.html#withTransferFunction)
    ///
    /// ``` self: ?*C.QColorSpace, transferFunction: qcolorspace_enums.TransferFunction, gamma: f32 ```
    pub fn WithTransferFunction2(self: ?*anyopaque, transferFunction: i64, gamma: f32) ?*C.QColorSpace {
        return C.QColorSpace_WithTransferFunction2(@ptrCast(self), @intCast(transferFunction), @floatCast(gamma));
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: ?*C.QColorSpace ```
    pub fn QDelete(self: ?*anyopaque) void {
        C.QColorSpace_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qcolorspace.html#types
pub const enums = struct {
    pub const NamedColorSpace = enum {
        pub const SRgb: i32 = 1;
        pub const SRgbLinear: i32 = 2;
        pub const AdobeRgb: i32 = 3;
        pub const DisplayP3: i32 = 4;
        pub const ProPhotoRgb: i32 = 5;
    };

    pub const Primaries = enum {
        pub const Custom: i32 = 0;
        pub const SRgb: i32 = 1;
        pub const AdobeRgb: i32 = 2;
        pub const DciP3D65: i32 = 3;
        pub const ProPhotoRgb: i32 = 4;
    };

    pub const TransferFunction = enum {
        pub const Custom: i32 = 0;
        pub const Linear: i32 = 1;
        pub const Gamma: i32 = 2;
        pub const SRgb: i32 = 3;
        pub const ProPhotoRgb: i32 = 4;
    };
};

File: src_original/libqcolorspace.cpp
#include <QByteArray>
#include <QColorSpace>
#include <QColorTransform>
#include <QList>
#include <QPointF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qcolorspace.h>
#include "libqcolorspace.h"
#include "libqcolorspace.hxx"

QColorSpace* QColorSpace_new() {
    return new QColorSpace();
}

QColorSpace* QColorSpace_new2(int namedColorSpace) {
    return new QColorSpace(static_cast<QColorSpace::NamedColorSpace>(namedColorSpace));
}

QColorSpace* QColorSpace_new3(QPointF* whitePoint, int transferFunction) {
    return new QColorSpace(*whitePoint, static_cast<QColorSpace::TransferFunction>(transferFunction));
}

QColorSpace* QColorSpace_new4(QPointF* whitePoint, const libqt_list /* of uint16_t */ transferFunctionTable) {
    QList<uint16_t> transferFunctionTable_QList;
    transferFunctionTable_QList.reserve(transferFunctionTable.len);
    uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
    for (size_t i = 0; i < transferFunctionTable.len; ++i) {
        transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
    }
    return new QColorSpace(*whitePoint, transferFunctionTable_QList);
}

QColorSpace* QColorSpace_new5(int primaries, int transferFunction) {
    return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), static_cast<QColorSpace::TransferFunction>(transferFunction));
}

QColorSpace* QColorSpace_new6(int primaries, float gamma) {
    return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), static_cast<float>(gamma));
}

QColorSpace* QColorSpace_new7(int primaries, const libqt_list /* of uint16_t */ transferFunctionTable) {
    QList<uint16_t> transferFunctionTable_QList;
    transferFunctionTable_QList.reserve(transferFunctionTable.len);
    uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
    for (size_t i = 0; i < transferFunctionTable.len; ++i) {
        transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
    }
    return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), transferFunctionTable_QList);
}

QColorSpace* QColorSpace_new8(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, int transferFunction) {
    return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, static_cast<QColorSpace::TransferFunction>(transferFunction));
}

QColorSpace* QColorSpace_new9(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, const libqt_list /* of uint16_t */ transferFunctionTable) {
    QList<uint16_t> transferFunctionTable_QList;
    transferFunctionTable_QList.reserve(transferFunctionTable.len);
    uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
    for (size_t i = 0; i < transferFunctionTable.len; ++i) {
        transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
    }
    return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, transferFunctionTable_QList);
}

QColorSpace* QColorSpace_new10(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, const libqt_list /* of uint16_t */ redTransferFunctionTable, const libqt_list /* of uint16_t */ greenTransferFunctionTable, const libqt_list /* of uint16_t */ blueTransferFunctionTable) {
    QList<uint16_t> redTransferFunctionTable_QList;
    redTransferFunctionTable_QList.reserve(redTransferFunctionTable.len);
    uint16_t* redTransferFunctionTable_arr = static_cast<uint16_t*>(redTransferFunctionTable.data);
    for (size_t i = 0; i < redTransferFunctionTable.len; ++i) {
        redTransferFunctionTable_QList.push_back(static_cast<uint16_t>(redTransferFunctionTable_arr[i]));
    }
    QList<uint16_t> greenTransferFunctionTable_QList;
    greenTransferFunctionTable_QList.reserve(greenTransferFunctionTable.len);
    uint16_t* greenTransferFunctionTable_arr = static_cast<uint16_t*>(greenTransferFunctionTable.data);
    for (size_t i = 0; i < greenTransferFunctionTable.len; ++i) {
        greenTransferFunctionTable_QList.push_back(static_cast<uint16_t>(greenTransferFunctionTable_arr[i]));
    }
    QList<uint16_t> blueTransferFunctionTable_QList;
    blueTransferFunctionTable_QList.reserve(blueTransferFunctionTable.len);
    uint16_t* blueTransferFunctionTable_arr = static_cast<uint16_t*>(blueTransferFunctionTable.data);
    for (size_t i = 0; i < blueTransferFunctionTable.len; ++i) {
        blueTransferFunctionTable_QList.push_back(static_cast<uint16_t>(blueTransferFunctionTable_arr[i]));
    }
    return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, redTransferFunctionTable_QList, greenTransferFunctionTable_QList, blueTransferFunctionTable_QList);
}

QColorSpace* QColorSpace_new11(const QColorSpace* colorSpace) {
    return new QColorSpace(*colorSpace);
}

QColorSpace* QColorSpace_new12(QPointF* whitePoint, int transferFunction, float gamma) {
    return new QColorSpace(*whitePoint, static_cast<QColorSpace::TransferFunction>(transferFunction), static_cast<float>(gamma));
}

QColorSpace* QColorSpace_new13(int primaries, int transferFunction, float gamma) {
    return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), static_cast<QColorSpace::TransferFunction>(transferFunction), static_cast<float>(gamma));
}

QColorSpace* QColorSpace_new14(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, int transferFunction, float gamma) {
    return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, static_cast<QColorSpace::TransferFunction>(transferFunction), static_cast<float>(gamma));
}

void QColorSpace_OperatorAssign(QColorSpace* self, const QColorSpace* colorSpace) {
    self->operator=(*colorSpace);
}

void QColorSpace_Swap(QColorSpace* self, QColorSpace* colorSpace) {
    self->swap(*colorSpace);
}

int QColorSpace_Primaries(const QColorSpace* self) {
    return static_cast<int>(self->primaries());
}

int QColorSpace_TransferFunction(const QColorSpace* self) {
    return static_cast<int>(self->transferFunction());
}

float QColorSpace_Gamma(const QColorSpace* self) {
    return self->gamma();
}

libqt_string QColorSpace_Description(const QColorSpace* self) {
    QString _ret = self->description();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

void QColorSpace_SetDescription(QColorSpace* self, const libqt_string description) {
    QString description_QString = QString::fromUtf8(description.data, description.len);
    self->setDescription(description_QString);
}

void QColorSpace_SetTransferFunction(QColorSpace* self, int transferFunction) {
    self->setTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction));
}

void QColorSpace_SetTransferFunction2(QColorSpace* self, const libqt_list /* of uint16_t */ transferFunctionTable) {
    QList<uint16_t> transferFunctionTable_QList;
    transferFunctionTable_QList.reserve(transferFunctionTable.len);
    uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
    for (size_t i = 0; i < transferFunctionTable.len; ++i) {
        transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
    }
    self->setTransferFunction(transferFunctionTable_QList);
}

void QColorSpace_SetTransferFunctions(QColorSpace* self, const libqt_list /* of uint16_t */ redTransferFunctionTable, const libqt_list /* of uint16_t */ greenTransferFunctionTable, const libqt_list /* of uint16_t */ blueTransferFunctionTable) {
    QList<uint16_t> redTransferFunctionTable_QList;
    redTransferFunctionTable_QList.reserve(redTransferFunctionTable.len);
    uint16_t* redTransferFunctionTable_arr = static_cast<uint16_t*>(redTransferFunctionTable.data);
    for (size_t i = 0; i < redTransferFunctionTable.len; ++i) {
        redTransferFunctionTable_QList.push_back(static_cast<uint16_t>(redTransferFunctionTable_arr[i]));
    }
    QList<uint16_t> greenTransferFunctionTable_QList;
    greenTransferFunctionTable_QList.reserve(greenTransferFunctionTable.len);
    uint16_t* greenTransferFunctionTable_arr = static_cast<uint16_t*>(greenTransferFunctionTable.data);
    for (size_t i = 0; i < greenTransferFunctionTable.len; ++i) {
        greenTransferFunctionTable_QList.push_back(static_cast<uint16_t>(greenTransferFunctionTable_arr[i]));
    }
    QList<uint16_t> blueTransferFunctionTable_QList;
    blueTransferFunctionTable_QList.reserve(blueTransferFunctionTable.len);
    uint16_t* blueTransferFunctionTable_arr = static_cast<uint16_t*>(blueTransferFunctionTable.data);
    for (size_t i = 0; i < blueTransferFunctionTable.len; ++i) {
        blueTransferFunctionTable_QList.push_back(static_cast<uint16_t>(blueTransferFunctionTable_arr[i]));
    }
    self->setTransferFunctions(redTransferFunctionTable_QList, greenTransferFunctionTable_QList, blueTransferFunctionTable_QList);
}

QColorSpace* QColorSpace_WithTransferFunction(const QColorSpace* self, int transferFunction) {
    return new QColorSpace(self->withTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction)));
}

QColorSpace* QColorSpace_WithTransferFunction2(const QColorSpace* self, const libqt_list /* of uint16_t */ transferFunctionTable) {
    QList<uint16_t> transferFunctionTable_QList;
    transferFunctionTable_QList.reserve(transferFunctionTable.len);
    uint16_t* transferFunctionTable_arr = static_cast<uint16_t*>(transferFunctionTable.data);
    for (size_t i = 0; i < transferFunctionTable.len; ++i) {
        transferFunctionTable_QList.push_back(static_cast<uint16_t>(transferFunctionTable_arr[i]));
    }
    return new QColorSpace(self->withTransferFunction(transferFunctionTable_QList));
}

QColorSpace* QColorSpace_WithTransferFunctions(const QColorSpace* self, const libqt_list /* of uint16_t */ redTransferFunctionTable, const libqt_list /* of uint16_t */ greenTransferFunctionTable, const libqt_list /* of uint16_t */ blueTransferFunctionTable) {
    QList<uint16_t> redTransferFunctionTable_QList;
    redTransferFunctionTable_QList.reserve(redTransferFunctionTable.len);
    uint16_t* redTransferFunctionTable_arr = static_cast<uint16_t*>(redTransferFunctionTable.data);
    for (size_t i = 0; i < redTransferFunctionTable.len; ++i) {
        redTransferFunctionTable_QList.push_back(static_cast<uint16_t>(redTransferFunctionTable_arr[i]));
    }
    QList<uint16_t> greenTransferFunctionTable_QList;
    greenTransferFunctionTable_QList.reserve(greenTransferFunctionTable.len);
    uint16_t* greenTransferFunctionTable_arr = static_cast<uint16_t*>(greenTransferFunctionTable.data);
    for (size_t i = 0; i < greenTransferFunctionTable.len; ++i) {
        greenTransferFunctionTable_QList.push_back(static_cast<uint16_t>(greenTransferFunctionTable_arr[i]));
    }
    QList<uint16_t> blueTransferFunctionTable_QList;
    blueTransferFunctionTable_QList.reserve(blueTransferFunctionTable.len);
    uint16_t* blueTransferFunctionTable_arr = static_cast<uint16_t*>(blueTransferFunctionTable.data);
    for (size_t i = 0; i < blueTransferFunctionTable.len; ++i) {
        blueTransferFunctionTable_QList.push_back(static_cast<uint16_t>(blueTransferFunctionTable_arr[i]));
    }
    return new QColorSpace(self->withTransferFunctions(redTransferFunctionTable_QList, greenTransferFunctionTable_QList, blueTransferFunctionTable_QList));
}

void QColorSpace_SetPrimaries(QColorSpace* self, int primariesId) {
    self->setPrimaries(static_cast<QColorSpace::Primaries>(primariesId));
}

void QColorSpace_SetPrimaries2(QColorSpace* self, const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint) {
    self->setPrimaries(*whitePoint, *redPoint, *greenPoint, *bluePoint);
}

void QColorSpace_SetWhitePoint(QColorSpace* self, QPointF* whitePoint) {
    self->setWhitePoint(*whitePoint);
}

QPointF* QColorSpace_WhitePoint(const QColorSpace* self) {
    return new QPointF(self->whitePoint());
}

uint8_t QColorSpace_TransformModel(const QColorSpace* self) {
    return static_cast<uint8_t>(self->transformModel());
}

uint8_t QColorSpace_ColorModel(const QColorSpace* self) {
    return static_cast<uint8_t>(self->colorModel());
}

void QColorSpace_Detach(QColorSpace* self) {
    self->detach();
}

bool QColorSpace_IsValid(const QColorSpace* self) {
    return self->isValid();
}

bool QColorSpace_IsValidTarget(const QColorSpace* self) {
    return self->isValidTarget();
}

QColorSpace* QColorSpace_FromIccProfile(const libqt_string iccProfile) {
    QByteArray iccProfile_QByteArray(iccProfile.data, iccProfile.len);
    return new QColorSpace(QColorSpace::fromIccProfile(iccProfile_QByteArray));
}

libqt_string QColorSpace_IccProfile(const QColorSpace* self) {
    QByteArray _qb = self->iccProfile();
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

QColorTransform* QColorSpace_TransformationToColorSpace(const QColorSpace* self, const QColorSpace* colorspace) {
    return new QColorTransform(self->transformationToColorSpace(*colorspace));
}

QVariant* QColorSpace_ToQVariant(const QColorSpace* self) {
    return new QVariant(self->operator QVariant());
}

void QColorSpace_SetTransferFunction22(QColorSpace* self, int transferFunction, float gamma) {
    self->setTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction), static_cast<float>(gamma));
}

QColorSpace* QColorSpace_WithTransferFunction22(const QColorSpace* self, int transferFunction, float gamma) {
    return new QColorSpace(self->withTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction), static_cast<float>(gamma)));
}

void QColorSpace_Delete(QColorSpace* self) {
    delete self;
}
